README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# meedr <img src='man/figures/logo.png' align="right" height="139" />

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/meedr)](https://CRAN.R-project.org/package=meedr)
[![R-CMD-check](https://github.com/schoulten/meedr/workflows/R-CMD-check/badge.svg)](https://github.com/schoulten/meedr/actions)
[![CRAN
downloads](http://cranlogs.r-pkg.org/badges/grand-total/meedr?color=green)](https://cran.r-project.org/package=meedr)
<!-- badges: end -->

The goal of **meedr** is to provide quick and easy access to market
expectations data to the main macroeconomic indicators in the Focus
report, made available by the **Central Bank of Brazil** through the
Expectations System data [API](https://dadosabertos.bcb.gov.br/). This
data comes from several financial institutions, such as: banks, brokers,
funds, consultancies, etc.

**Warning**: This package was definitively archived by CRAN on July 22,
2021 and is only available in this repository. Despite this, maintenance
and bug fixes will continue to be done.

The **meedr** package offers an R interface to the API and other
advantages:

-   Use of a caching system with package `memoise` to speed up repeated
    requests of data;
-   User can utilize all cores of the machine (parallel computing) when
    fetching a large batch of time series.

Check the [meedr pkgdown page](https://fortietwo.com/meedr/) for more
general information ;)

## Installation

You can install the development version from
[GitHub](https://github.com/schoulten/meedr) with:

``` r
# install.packages("remotes")
remotes::install_github("schoulten/meedr")
```

## Features

-   [get_monthly()](#get_monthly): Get data on monthly market
    expectations
-   [get_quarterly()](#get_quarterly): Get data on quarterly market
    expectations
-   [get_annual()](#get_annual): Get data on annual market expectations
-   [get_inflation_12m()](#get_inflation_12m): Get data on market
    expectations for inflation over the next 12 months
-   [get_monthly_top5()](#get_monthly_top5): Get data on monthly market
    expectations for the Top 5 indicators
-   [get_annual_top5()](#get_annual_top5): Get data on annual market
    expectations for the Top 5 indicators
-   [get_selic()](#get_selic): Get Selic market expectations data

## Example

These are some basic examples of using the package:

### get_monthly()

``` r
library(meedr)

# Monthly market expectations for IPCA indicator
ipca <- meedr::get_monthly(
  indicator      = "IPCA",
  first_date     = Sys.Date()-30,
  reference_date = format(Sys.Date(), "%m/%Y"),
  be_quiet       = TRUE
  )

head(ipca, 5)
#> # A tibble: 5 × 10
#>   indicator date       reference_date   mean median    sd    min   max
#>   <chr>     <date>     <chr>           <dbl>  <dbl> <dbl>  <dbl> <dbl>
#> 1 IPCA      2022-08-05 08/2022        -0.149 -0.15  0.180 -0.580  0.35
#> 2 IPCA      2022-08-05 08/2022        -0.125 -0.15  0.197 -0.580  0.55
#> 3 IPCA      2022-08-04 08/2022        -0.157 -0.17  0.181 -0.840  0.35
#> 4 IPCA      2022-08-04 08/2022        -0.117 -0.135 0.207 -0.840  0.55
#> 5 IPCA      2022-08-03 08/2022        -0.154 -0.165 0.182 -0.840  0.35
#> # … with 2 more variables: n_respondents <int>, basis <int>
```

### get_quarterly()

``` r
# Quarterly market expectations for GDP indicator
meedr::get_quarterly(
  indicator      = "PIB Total",
  first_date     = "2021-01-01",
  reference_date = paste0(lubridate::quarter(Sys.Date()), "/", lubridate::year(Sys.Date())),
  be_quiet       = TRUE
  )
```

### get_annual()

``` r
# Annual market expectations for SELIC and exchange rate (BRL) indicator
meedr::get_annual(
  indicator      = c("Selic", "Câmbio"),
  reference_date = format(Sys.Date(), "%Y"),
  be_quiet       = TRUE
  )
```

### get_inflation_12m()

``` r
# Inflation over the next 12 months
# First, and a suggestion, run this for using parallel computing:
future::plan(future::multisession, workers = floor(future::availableCores()/2))
meedr::get_inflation_12m(
  indicator   = c("IGP-DI", "IGP-M", "INPC", "IPA-DI", "IPA-M", "IPCA", "IPCA-15", "IPC-Fipe"),
  smoothed    = "yes",
  be_quiet    = FALSE, # display messages
  do_parallel = TRUE # turn on parallel computing
  )
```

### get_monthly_top5()

``` r
# Monthly market expectations for IGP-M indicator (Top 5 Focus)
meedr::get_monthly_top5(
  indicator  = "IGP-M",
  first_date = NULL, # get all data to current date
  calc_type  = "long",
  be_quiet   = TRUE
  )
```

### get_annual_top5()

``` r
# Annual market expectations for SELIC indicator (Top 5 Focus)
meedr::get_annual_top5(
  indicator   = "Selic",
  detail      = "Fim do ano", # argument deprecated by API BCB-Olinda
  be_quiet    = TRUE,
  use_memoise = FALSE # disable caching system
  )
```

### get_selic()

``` r
df1 <- get_selic(indicator = "Selic", first_date = Sys.Date() - 30)
#> 
#> Fetching [Selic] data from BCB-Olinda...
#> 
#> Found 576 observations!
df1
#> # A tibble: 576 × 10
#>    indicator date       meeting  mean median    sd   min   max n_respondents
#>    <chr>     <date>     <chr>   <dbl>  <dbl> <dbl> <dbl> <dbl>         <int>
#>  1 Selic     2022-08-05 R5/2024  8.91    9   0.854   7    10              14
#>  2 Selic     2022-08-05 R5/2024  8.91    9   0.854   7    10              14
#>  3 Selic     2022-08-05 R4/2024  8.87    9   0.894   7    10.5            30
#>  4 Selic     2022-08-05 R4/2024  9.12    9   0.987   7    12.2           107
#>  5 Selic     2022-08-05 R3/2024  9.34    9.5 0.949   7.5  11              33
#>  6 Selic     2022-08-05 R3/2024  9.52    9.5 0.991   7.5  12.5           112
#>  7 Selic     2022-08-05 R2/2024  9.81   10   0.894   8    11.5            33
#>  8 Selic     2022-08-05 R2/2024  9.97   10   0.980   7.5  12.8           112
#>  9 Selic     2022-08-05 R1/2024 10.4    10.4 0.934   8.5  13              34
#> 10 Selic     2022-08-05 R1/2024 10.4    10.5 0.968   8.5  13             113
#> # … with 566 more rows, and 1 more variable: basis <int>

df2 <- get_selic(indicator = "Selic", first_date = NULL, meeting = "R1/2024")
#> 
#> Fetching [Selic] data from BCB-Olinda...
#> 
#> Found 254 observations!
df2
#> # A tibble: 254 × 10
#>    indicator date       meeting  mean median    sd   min   max n_respondents
#>    <chr>     <date>     <chr>   <dbl>  <dbl> <dbl> <dbl> <dbl>         <int>
#>  1 Selic     2022-08-05 R1/2024  10.4   10.4 0.934  8.5   13              34
#>  2 Selic     2022-08-05 R1/2024  10.4   10.5 0.968  8.5   13             113
#>  3 Selic     2022-08-04 R1/2024  10.5   10.5 0.933  8.5   12.5            63
#>  4 Selic     2022-08-04 R1/2024  10.4   10.5 0.954  8.5   13             113
#>  5 Selic     2022-08-03 R1/2024  10.5   10.5 1.04   7.25  13              64
#>  6 Selic     2022-08-03 R1/2024  10.4   10.5 0.990  7.25  13             113
#>  7 Selic     2022-08-02 R1/2024  10.5   10.5 1.06   7.25  13              67
#>  8 Selic     2022-08-02 R1/2024  10.4   10.5 0.990  7.25  13             113
#>  9 Selic     2022-08-01 R1/2024  10.5   10.5 1.07   7.25  13              70
#> 10 Selic     2022-08-01 R1/2024  10.4   10.5 0.989  7.25  13             113
#> # … with 244 more rows, and 1 more variable: basis <int>
```

## Related works

Check out some similar works:

-   [rbcb](https://github.com/wilsonfreitas/rbcb): R interface to
    Brazilian Central Bank web services by @wilsonfreitas


NEWS.md
# meedr 0.0.4
* New function: get_selic() collects data from the ExpectativasMercadoSelic feature of the Olinda API

# meedr 0.0.3
* Updates list of indicators due to API BCB-Olinda change

# meedr 0.0.2
* Adequate API request in accordance with CRAN policy

# meedr 0.0.1
* First version released on CRAN.


R/get_selic.R
#' Get Selic market expectations data (Focus/BCB)
#'
#' @param indicator A character vector with economic indicators names, see details for possible value. Defaults to "Selic".
#' @param first_date A single string or Date in "YYYY-mm-dd" format. Default for 2 years prior to the current date.
#' @param last_date A single string or Date in "YYYY-mm-dd" format. Default is the current date.
#' @param meeting A single string in "RX/YYYY" format, where X is a positive integer, indicating the meeting for which the statistic is expected. Defaults to NULL.
#' @param be_quiet Logical. Should messages or warnings not be displayed? Defaults to FALSE.
#' @param use_memoise Logical. Sets the use of caching system, creating a folder named "cache_bcb" by default. Defaults to FALSE.
#' @param do_parallel Logical. For using parallel data importation. Defaults to FALSE.
#'
#' @return A tibble with data from BCB-Olinda Data Services.
#' @description This function provides the extraction of data and statistics related to expectations of economic indicators, specifically Selic market expectations, made available by the Central Bank of Brazil's Expectations System through the Olinda API. The data comes from several financial institutions: banks, funds, research houses, etc. Important: arguments are case sensitive.
#' @details For periods for which there are no statistics, they will be omitted from the query.
#'
#' Possible values for indicator argument: 'Selic'.
#'
#' @author Fernando da Silva <<fernando@fortietwo.com>>
#' @encoding UTF-8
#' @export
#'
#' @examples
#' df1 <- get_selic(indicator = "Selic", first_date = Sys.Date() - 30)
#' df1
#'
#' df2 <- get_selic(indicator = "Selic", first_date = NULL, meeting = "R1/2024")
#' df2
get_selic <- function(
    indicator      = "Selic",
    first_date     = Sys.Date() - 2*365,
    last_date      = Sys.Date(),
    meeting        = NULL,
    be_quiet       = FALSE,
    use_memoise    = FALSE,
    do_parallel    = FALSE
    ) {

  valid_indicator <- c("Selic")

  if (missing(indicator) | !all(indicator %in% valid_indicator) | is.null(indicator)) {
    stop("\nArgument 'indicator' is not valid or missing. Check your inputs.", call. = FALSE)
  }

  first_date <- try(as.Date(first_date), silent = TRUE)
  if (length(first_date) <= 0 || is.na(first_date)) {first_date = NULL}
  if (inherits(first_date, "try-error")) {
    stop("\nArgument 'first_date' is not a valid date.", call. = FALSE)
  }
  if (missing(first_date)) {first_date = Sys.Date() - 10 * 365}

  last_date <- try(as.Date(last_date), silent = TRUE)
  if (length(last_date) <= 0 || is.na(last_date)) {last_date = NULL}
  if (inherits(last_date, "try-error")) {
    stop("\nArgument 'last_date' is not a valid date.", call. = FALSE)
  }
  if (missing(last_date)) {last_date = Sys.Date() - 10 * 365}

  if ((length(first_date) > 0) && first_date > Sys.Date()) {
    stop("\nIt seems that 'first_date' > current date. Check your inputs.", call. = FALSE)
  }

  if (!is.null(last_date)) {
    if ((length(first_date) > 0) && last_date < first_date) {
      stop("\nIt seems that 'last_date' < first_date. Check your inputs.", call. = FALSE)
    }
  }

  if (!is.null(meeting) && !is.na(meeting)) {
    if (!inherits(meeting, "character")) {
      stop("\nArgument 'meeting' is not valid. Check your inputs.", call. = FALSE)
    } else if
    (nchar(meeting) == 7L & (grepl("(R\\d{1})(\\/{1})(\\d{4}$)", meeting))) {
      meeting <- as.character(meeting)
    } else
      stop("\nArgument 'meeting' is not valid. Check your inputs.", call. = FALSE)
  } else if
  (is.na(meeting) && (length(meeting) > 0)) {
    meeting <- NULL
    } else meeting <- NULL

  if (!inherits(do_parallel, "logical") || (is.na(do_parallel))) {
    stop("\nArgument 'do_parallel' must be logical. Check your inputs.", call. = FALSE)
  }

  if (!inherits(be_quiet, "logical") || (is.na(be_quiet))) {
    stop("\nArgument 'be_quiet' must be logical. Check your inputs.", call. = FALSE)
  }

  foo_args <- paste0(
    paste0("(", paste(sprintf("Indicador eq '%s'", indicator), collapse = " or ", sep = ""), ")"),
    sprintf(" and Data ge '%s'", first_date),
    sprintf(" and Data le '%s'", last_date),
    sprintf(" and Reuniao eq '%s'", meeting)
    )

  odata_url <- list(
    httr::modify_url(
      "https://olinda.bcb.gov.br/olinda/servico/Expectativas/versao/v1/odata/ExpectativasMercadoSelic",
      query = list(
        `$filter`  = foo_args,
        `$format`  = "json",
        `$orderby` = "Data desc"
        )
      )
    )

  if (!inherits(use_memoise, "logical") || (is.na(use_memoise))) {
    stop("\nArgument 'use_memoise' must be logical. Check your inputs.", call. = FALSE)
  } else
    memoising <- function(use_memoise, cache_dir) {
      if (use_memoise) {
        foo_memoise <- memoise::memoise(f = jsonlite::fromJSON, cache = cache_dir)
      } else
        foo_memoise <- jsonlite::fromJSON
    }

  from_bcb <- memoising(
    use_memoise = use_memoise,
    cache_dir   = memoise::cache_filesystem("./cache_bcb")
    )

  resp <- httr::GET(odata_url[[1]])
  if (httr::http_type(resp) != "application/json") {
    stop("BCB-Olinda API did not return json.", call. = FALSE)
  }

  if (!do_parallel) {

    if (be_quiet) {message("", appendLF = FALSE)} else {
      message(
        paste0("\nFetching [", paste(indicator, collapse = ", "), "] data ", "from BCB-Olinda... \n"),
        appendLF = FALSE
      )
    }

    df <- try(
      suppressWarnings(purrr::pmap(.l = odata_url, .f = from_bcb)[[1]][["value"]]),
      silent = TRUE
    )
  } else {
    used_workers <- future::nbrOfWorkers()
    available_cores <- future::availableCores()
    flag <- inherits(future::plan(), "sequential")
    if (flag) {
      stop(paste0(
        "When using do_parallel = TRUE, you need to call future::plan() to configure your parallel settings.\n",
        "A suggestion, write the following lines just before:\n\n",
        "future::plan(future::multisession, workers = floor(future::availableCores()/2))", "\n\n",
        "Notice it will use half of your available cores so that your OS has some room to breathe."),
        call. = FALSE
      )
    } else if
    (be_quiet) {message("", appendLF = FALSE)} else
      message(
        paste0("\nRunning parallel with ", used_workers, " cores (", available_cores, " available)\n",
               "\nFetching [", paste(indicator, collapse = ", "), "] data ", "from BCB-Olinda... \n"),
        appendLF = TRUE
      )

    df <- try(
      suppressWarnings(furrr::future_pmap(.l = odata_url, .f = from_bcb)[[1]][["value"]]),
      silent = TRUE
    )
  }

  if (inherits(df, "try-error")) {
    stop(
      "\nProblem in fetching data, error message: ", conditionMessage(attr(df, "condition")),
      call. = FALSE
      )
  } else if (purrr::is_empty(df)) {
    stop(
      paste0(
        "\nIt seems that there is no data available. Possibly, the last available data is earlier than that defined in one of these arguments:
      \n1. 'first_date'", "\n2. 'meeting'"
      ),
      call. = FALSE
      )
  } else if (be_quiet) {
    message("", appendLF = FALSE)
  } else
    message(paste0("\nFound ", nrow(df), " observations!\n"), appendLF = FALSE)

  new_names <- c(
    "indicator"      = "Indicador",
    "detail"         = "IndicadorDetalhe",
    "date"           = "Data",
    "meeting"        = "Reuniao",
    "mean"           = "Media",
    "median"         = "Mediana",
    "sd"             = "DesvioPadrao",
    "coef_var"       = "CoeficienteVariacao",
    "min"            = "Minimo",
    "max"            = "Maximo",
    "n_respondents"  = "numeroRespondentes",
    "basis"          = "baseCalculo"
    )

  df <- dplyr::rename(
    dplyr::as_tibble(df),
    dplyr::any_of(new_names)
    )

  df <- dplyr::mutate(df, date = as.Date(date, format = "%Y-%m-%d"))

  return(df)

}


R/get_annual.R
#' Get data on annual market expectations for economic indicators (Focus/BCB)
#'
#' @param indicator A single string or a character vector with economic indicators names, see details for possible values. Defaults to NULL.
#' @param detail A single string. See details for possible values. Defaults to NULL.
#' @param first_date A single string or Date in "YYYY-mm-dd" format. Default for 2 years prior to the current date.
#' @param last_date A single string or Date in "YYYY-mm-dd". format Default is the current date.
#' @param reference_date A single string in "YYYY" format, indicating the reference year for which the statistic is expected. Defaults to NULL.
#' @param be_quiet Logical. Should messages or warnings not be displayed? Defaults to FALSE.
#' @param use_memoise Logical. Sets the use of caching system, creating a folder named "cache_bcb" by default. Default to TRUE.
#' @param do_parallel Logical. For using parallel data importation. Defaults to FALSE.
#'
#' @return A tibble with data from BCB-Olinda Data Services.
#' @description This function provides the extraction of data and statistics related to the expectations of economic indicators, specifically the annual market expectations, made available by the Central Bank of Brazil's Expectations System through the Olinda API. The data comes from several financial institutions: banks, funds, research houses, etc. Important: arguments are case sensitive.
#' @details For periods for which there are no statistics, they will be omitted from the query.
#'
#' Possible values for indicator argument: 'Balança comercial', 'Conta corrente', 'Investimento direto no país', 'Dívida bruta do governo geral', 'Dívida líquida do setor público', 'Resultado primário', 'Resultado nominal', 'IGP-DI', 'IGP-M', 'INPC', 'IPA-DI', 'IPA-M', 'IPCA', 'IPCA-15', 'IPC-Fipe', 'IPCA Administrados', 'IPCA Alimentação no domicílio', 'IPCA Bens industrializados', 'IPCA Livres', 'IPCA Serviços', 'Produção industrial', 'PIB Agropecuária', 'PIB Indústria', 'PIB Serviços', 'PIB Total', 'PIB Despesa de consumo da administração pública', 'PIB Despesa de consumo das famílias', 'PIB Exportação de bens e serviços', 'PIB Formação Bruta de Capital Fixo', 'PIB Importação de bens e serviços', 'Taxa de desocupação', 'Selic', 'Câmbio'.
#'
#' Possible values for detail argument: if indicator 'Balança comercial' -> 'Exportações', 'Importações', 'Saldo'.
#'
#' @author Fernando da Silva <<fernando@fortietwo.com>>
#' @encoding UTF-8
#' @export
#'
#' @examples
#' df <- get_annual(
#'   indicator = "IPCA",
#'   first_date = Sys.Date() - 30,
#'   reference_date = format(Sys.Date(), "%Y"),
#'   use_memoise = FALSE
#' )
get_annual <- function(
  indicator      = NULL,
  detail         = NULL,
  first_date     = Sys.Date() - 2*365,
  last_date      = Sys.Date(),
  reference_date = NULL,
  be_quiet       = FALSE,
  use_memoise    = TRUE,
  do_parallel    = FALSE
  ){
  valid_indicator <- c(
    # Externo
    "Balan\u00e7a comercial",
    "Conta corrente",
    "Investimento direto no pa\u00eds",
    # Fiscal
    "D\u00edvida bruta do governo geral",
    "D\u00edvida l\u00edquida do setor p\u00fablico",
    "Resultado prim\u00e1rio",
    "Resultado nominal",
    # Índices de preços
    "IGP-DI",
    "IGP-M",
    "INPC",
    "IPA-DI",
    "IPA-M",
    "IPCA",
    "IPCA-15",
    "IPC-Fipe",
    "IPCA Administrados",
    "IPCA Alimenta\u00e7\u00e3o no domic\u00edlio",
    "IPCA Bens industrializados",
    "IPCA Livres",
    "IPCA Servi\u00e7os",
    # Atividade
    "Produ\u00e7\u00e3o industrial",
    "PIB Agropecu\u00e1ria",
    "PIB Ind\u00fastria",
    "PIB Servi\u00e7os",
    "PIB Total",
    "PIB Despesa de consumo da administra\u00e7\u00e3o p\u00fablica",
    "PIB Despesa de consumo das fam\u00edlias",
    "PIB Exporta\u00e7\u00e3o de bens e servi\u00e7os",
    "PIB Forma\u00e7\u00e3o Bruta de Capital Fixo",
    "PIB Importa\u00e7\u00e3o de bens e servi\u00e7os",
    "Taxa de desocupa\u00e7\u00e3o",
    # Taxas
    "Selic",
    "C\u00e2mbio"
    )

  if (missing(indicator) | !all(indicator %in% valid_indicator) | is.null(indicator)) {
    stop("\nArgument 'indicator' is not valid or missing. Check your inputs.", call. = FALSE)
  }

  valid_detail <- c(
    "Balan\u00e7a comercial / Exporta\u00e7\u00f5es",
    "Balan\u00e7a comercial / Importa\u00e7\u00f5es",
    "Balan\u00e7a comercial / Saldo"
    )

  if (!is.null(detail) && !is.na(detail)) {
    if (!inherits(detail, "character")) {
      stop("\nArgument 'detail' is not valid. Check your inputs.", call. = FALSE)
    } else if
    (!all(paste0(indicator, " / ", detail) %in% valid_detail)) {
      stop("\nArgument 'detail' is not valid. Check your inputs.", call. = FALSE)
    } else if
    (length(detail) > 1) {
      stop("\nArgument 'detail' is not valid. Check your inputs.", call. = FALSE)
    } else
      detail
  } else if
  ((length(detail) > 0) && is.na(detail)) {
    detail <- NULL
  }

  first_date <- try(as.Date(first_date), silent = TRUE)
  if (length(first_date) <= 0 || is.na(first_date)) {first_date = NULL}
  if (inherits(first_date, "try-error")) {
    stop("\nArgument 'first_date' is not a valid date.", call. = FALSE)
  }
  if (missing(first_date)) {first_date = Sys.Date() - 10 * 365}

  last_date <- try(as.Date(last_date), silent = TRUE)
  if (length(last_date) <= 0 || is.na(last_date)) {last_date = NULL}
  if (inherits(last_date, "try-error")) {
    stop("\nArgument 'last_date' is not a valid date.", call. = FALSE)
  }
  if (missing(last_date)) {last_date = Sys.Date() - 10 * 365}

  if ((length(first_date) > 0) && first_date > Sys.Date()) {
    stop("\nIt seems that 'first_date' > current date. Check your inputs.", call. = FALSE)
  }

  if (!is.null(last_date)) {
    if ((length(first_date) > 0) && last_date < first_date) {
      stop("\nIt seems that 'last_date' < first_date. Check your inputs.", call. = FALSE)
    }
  }

  if (!is.null(reference_date) && !is.na(reference_date)) {
    if (!inherits(reference_date, "character")) {
      stop("\nArgument 'reference_date' is not valid. Check your inputs.", call. = FALSE)
    } else if
    (nchar(reference_date) == 4L & grepl("[[:digit:]]+$", reference_date)) {
      reference_date <- as.character(reference_date)
    } else
      stop("\nArgument 'reference_date' is not valid. Check yout inputs.", call. = FALSE)
  } else if
  (is.na(reference_date) && (length(reference_date) > 0)) {reference_date <- NULL}

  if (!inherits(do_parallel, "logical") || (is.na(do_parallel))) {
    stop("\nArgument 'do_parallel' must be logical. Check your inputs.", call. = FALSE)
  } else if

  (!inherits(be_quiet, "logical") || (is.na(be_quiet))) {
    stop("\nArgument 'be_quiet' must be logical. Check your inputs.", call. = FALSE)
  }

  foo_args <- paste0(
    paste0("(", paste(sprintf("Indicador eq '%s'", indicator), collapse = " or ", sep = ""), ")"),
    sprintf(" and IndicadorDetalhe eq '%s'", detail),
    sprintf(" and Data ge '%s'", first_date),
    sprintf(" and Data le '%s'", last_date),
    sprintf(" and DataReferencia eq '%s'", reference_date)
  )

  odata_url <- list(
    httr::modify_url(
      "https://olinda.bcb.gov.br/olinda/servico/Expectativas/versao/v1/odata/ExpectativasMercadoAnuais",
      query = list(
        `$filter`  = foo_args,
        `$format`  = "json",
        `$orderby` = "Data desc"
      )
    )
  )

  if (!inherits(use_memoise, "logical") || (is.na(use_memoise))) {
    stop("\nArgument 'use_memoise' must be logical. Check your inputs.", call. = FALSE)
  } else
    memoising <- function(use_memoise, cache_dir) {
      if (use_memoise) {
        foo_memoise <- memoise::memoise(f = jsonlite::fromJSON, cache = cache_dir)
      } else
        foo_memoise <- jsonlite::fromJSON
    }

  from_bcb <- memoising(
    use_memoise = use_memoise,
    cache_dir   = memoise::cache_filesystem("./cache_bcb")
  )

  resp <- httr::GET(odata_url[[1]])
  if (httr::http_type(resp) != "application/json") {
    stop("BCB-Olinda API did not return json.", call. = FALSE)
  }

  if (!do_parallel) {

    if (be_quiet) {message("", appendLF = FALSE)} else {
      message(
        paste0("\nFetching [", paste(indicator, collapse = ", "), "] data ", "from BCB-Olinda... \n"),
        appendLF = FALSE
      )
    }

    df <- try(
      suppressWarnings(purrr::pmap(.l = odata_url, .f = from_bcb)[[1]][["value"]]),
      silent = TRUE
    )
  } else {
    used_workers <- future::nbrOfWorkers()
    available_cores <- future::availableCores()
    flag <- inherits(future::plan(), "sequential")
    if (flag) {
      stop(paste0(
        "When using do_parallel = TRUE, you need to call future::plan() to configure your parallel settings.\n",
        "A suggestion, write the following lines just before:\n\n",
        "future::plan(future::multisession, workers = floor(future::availableCores()/2))", "\n\n",
        "Notice it will use half of your available cores so that your OS has some room to breathe."),
        call. = FALSE
      )
    } else if
    (be_quiet) {message("", appendLF = FALSE)} else
      message(
        paste0("\nRunning parallel with ", used_workers, " cores (", available_cores, " available)\n",
               "\nFetching [", paste(indicator, collapse = ", "), "] data ", "from BCB-Olinda... \n"),
        appendLF = TRUE
      )

    df <- try(
      suppressWarnings(furrr::future_pmap(.l = odata_url, .f = from_bcb)[[1]][["value"]]),
      silent = TRUE
    )
  }

  if (inherits(df, "try-error")) {
    stop("\nProblem in fetching data, error message: ", conditionMessage(attr(df, "condition")),
         call. = FALSE
    )
  } else if
  (purrr::is_empty(df)) {
    stop(
      paste0(
        "\nIt seems that there is no data available. Possibly, the last available data is earlier than that defined in one of these arguments:
      \n1. 'first_date'", "\n2. 'reference_date'"
      ),
      call. = FALSE
    )
  } else if
  (be_quiet) {message("", appendLF = FALSE)}
  else
    message(paste0("\nFound ", nrow(df), " observations!\n"), appendLF = FALSE)

  new_names <- c(
    "indicator"      = "Indicador",
    "detail"         = "IndicadorDetalhe",
    "date"           = "Data",
    "reference_date" = "DataReferencia",
    "mean"           = "Media",
    "median"         = "Mediana",
    "sd"             = "DesvioPadrao",
    "coef_var"       = "CoeficienteVariacao",
    "min"            = "Minimo",
    "max"            = "Maximo",
    "n_respondents"  = "numeroRespondentes",
    "basis"          = "baseCalculo"
    )

  df <- dplyr::rename(
    dplyr::as_tibble(df),
    dplyr::any_of(new_names)
    )

  df <- dplyr::mutate(df, date = as.Date(date, format = "%Y-%m-%d"))

  return(df)
}
